Expose triggering on-demand full snapshot via HTTP endpoint

shreyas-s-rao opened the pull request:
**What this PR does / why we need it**:
This PR adds functionality to trigger on-demand full snapshots via the HTTP endpoint `/snapshot/full`.
This PR also fixes a few unhandled scenarios for the case when snapstore provider is not configured, such as etcd readiness probe handling and defragmentation. It also refactors the server command to make it more modular.

**Which issue(s) this PR fixes**:
Fixes #113 

**Special notes for your reviewer**:
Also refactored the `TriggerFullSnapshot` and `DefragDataPeriodically` methods, as well as the order of object creation in `NewServerCommand` in order to pass the `Snapshotter` to the `HTTPHandler`.
Will add unit tests shortly.

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy
- target_group:   user|operator
-->
```noteworthy user
Added functionality to trigger on-demand full snapshots via the HTTP endpoint `/snapshot/full`.
```

swapnilgm commented on pkg/server/httpAPI.go:
```diff
@@ -182,3 +185,12 @@ func (h *HTTPHandler) serveInitializationStatus(rw http.ResponseWriter, req *htt
+
+// serveSnapshotFull triggers an out-of-schedule full snapshot
+// for the configured Snapshotter
+func (h *HTTPHandler) serveSnapshotFull(rw http.ResponseWriter, req *http.Request) {
```
Rename as `serveFullSnapshotTrigger`.

shreyas-s-rao replied:
Done.

swapnilgm commented on pkg/server/httpAPI.go:
```diff
@@ -182,3 +185,12 @@ func (h *HTTPHandler) serveInitializationStatus(rw http.ResponseWriter, req *htt
+// for the configured Snapshotter
+func (h *HTTPHandler) serveSnapshotFull(rw http.ResponseWriter, req *http.Request) {
+	if err := h.Snapshotter.TriggerFullSnapshot(); err != nil {
+		h.Logger.Warnf("skipped triggering out-of-schedule full snapshot: %v", err)
```
Return http status code 405 on error.

shreyas-s-rao replied:
Done.

swapnilgm commented on cmd/server.go:
```diff
@@ -67,43 +67,17 @@ func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
-				ReqCh:           make(chan struct{}),
-				AckCh:           make(chan struct{}),
+			if storageProvider == "" {
+				logger.Warnf("No snapstore storage provider configured. Will not start backup schedule. Terminating !!")
```
We can't terminate here. We wan't `handler` to be up and running for on demand  initialisation purpose, specifically etcd-events case.

Keep the flow in file same as before. Just create `ssr` object before `handler` start if `storageProvider` is not empty; otherwise pass `nil` to handler. I guess not `handler` have `snapshotter` as well. You can refactor code to avoid `handleNoSsrRequest` call by adding `nil` check in `serverInitialize` function itself.

shreyas-s-rao replied:
Agreed. I have refactored the code to avoid `handleNoSsrRequest`.

shreyas-s-rao commented:
@swapnilgm I have made the suggested changes and also refactored code to wait for etcd probe to succeed before setting http status to OK, even in the case of etcd-events.

ashwani2k commented on pkg/server/httpAPI.go:
```diff
@@ -182,3 +189,19 @@ func (h *HTTPHandler) serveInitializationStatus(rw http.ResponseWriter, req *htt
+	}
+	if err := h.Snapshotter.TriggerFullSnapshot(); err != nil {
+		h.Logger.Warnf("Skipped triggering out-of-schedule full snapshot: %v", err)
+		rw.WriteHeader(http.StatusMethodNotAllowed)
```
Shouldn't this be StatusInternalServerError  instead as there is a snapshotted configured so the method is allowed but something went wrong while taking the Snapshot.

shreyas-s-rao replied:
Nice catch @ashwani2k . Thanks!

shreyas-s-rao commented:
@amshuman-kr I did consider this, but given it would create a lot of code duplication, I stuck to the current approach.

amshuman-kr commented:
> it would create a lot of code duplication
@shreyas-s-rao how about smaller funcs for the reusable code?

shreyas-s-rao commented:
> @shreyas-s-rao how about smaller funcs for the reusable code?
Makes sense @amshuman-kr . I'll make the necessary changes.

shreyas-s-rao commented:
@amshuman-kr @swapnilgm I have addressed the suggestions. Also fixed a couple of things wrt no-snapstore-provider case, namely probe loop and periodic defrag (for etcd-events case).

amshuman-kr commented on cmd/server.go:
```diff
@@ -97,142 +143,147 @@ func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
+		if err != nil {
+			logger.Errorf("Failed to probe etcd: %v", err)
+			handler.Status = http.StatusServiceUnavailable
+			continue
```
Same as [above](https://github.com/gardener/etcd-backup-restore/pull/143/files#diff-1427b112aa72f64adcf839007e954d56R196).

shreyas-s-rao commented on cmd/server.go:
```diff
@@ -78,12 +85,51 @@ func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
+				insecureSkipVerify,
+				etcdEndpoints)
+
+			if snapshotterEnabled {
```
Addressed.

shreyas-s-rao commented on cmd/server.go:
```diff
@@ -97,142 +143,147 @@ func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
+		if err != nil {
+			logger.Errorf("Failed to probe etcd: %v", err)
+			handler.Status = http.StatusServiceUnavailable
+			continue
```
Yes Amshu. It would be better to quickly merge this PR and take this up in a separate PR.

swapnilgm replied:
The call timeout for etcd probe is set to default 30 sec i.e. each every probe call blocks for 30 sec if etcd is unavailable. I think its not necessary to wait in between as it dosen't introduce much network traffic as well.

shreyas-s-rao commented:
@amshuman-kr @swapnilgm I've addressed the latest suggestions. PTAL.

swapnilgm commented on pkg/server/httpAPI.go:
```diff
@@ -143,6 +146,16 @@ func (h *HTTPHandler) serveInitialize(rw http.ResponseWriter, req *http.Request)
+				// This is needed to stop the currently running snapshotter.
+				atomic.StoreUint32(&h.AckState, HandlerAckWaiting)
+				h.Logger.Info("Changed handler state.")
+				h.ReqCh <- emptyStruct
```
Why do we have two requests to stop snapshotter?

shreyas-s-rao commented on pkg/server/httpAPI.go:
```diff
@@ -143,6 +146,16 @@ func (h *HTTPHandler) serveInitialize(rw http.ResponseWriter, req *http.Request)
+				// This is needed to stop the currently running snapshotter.
+				atomic.StoreUint32(&h.AckState, HandlerAckWaiting)
+				h.Logger.Info("Changed handler state.")
+				h.ReqCh <- emptyStruct
```
Addressed. Thanks for the catch!
